lib/Rex/Helper/Rexfile/ParamLookup.pm

```diff
@@ -0,0 +1,103 @@
+#
+# (c) Jan Gehring <jan.gehring@gmail.com>
+#
+# vim: set ts=3 sw=3 tw=0:
+# vim: set expandtab:
+
+package Rex::Helper::Rexfile::ParamLookup;
+
+use strict;
+use warnings;
+
+# VERSION
+
+use Devel::Caller;
+use Data::Dumper;
+require Rex::Exporter;
+require Rex::Commands;
+
+use base qw(Rex::Exporter);
+use vars qw (@EXPORT);
+
+@EXPORT = qw(param_lookup);
+
+sub param_lookup {
+  my ( $key, $default ) = @_;
+
+  my $ret;
+
+  my ($caller_pkg) = caller(0);
+
+  my @args = Devel::Caller::caller_args(1);
+  if ( ref $args[0] eq "HASH" ) {
+    if ( exists $args[0]->{$key} ) {
+      $ret = $args[0]->{$key};
+    }
+  }
+
+  if ( !$ret ) {
+
+    # check if cmdb is loaded
+    my ($use_cmdb) = grep { m/CMDB\.pm/ } keys %INC;
+    if ($use_cmdb) {
+
+      # look inside cmdb
+      my $cmdb_key = "${caller_pkg}::$key";
+      $ret = Rex::Commands::get( Rex::CMDB::cmdb($cmdb_key) );
+
+      if ( !$ret ) {
+
+        # check in resource
+        if ($Rex::Resource::INSIDE_RES) {
+          $cmdb_key =
+              "${caller_pkg}::"
+            . $Rex::Resource::CURRENT_RES[-1]->display_name
+            . "::$key";
+          $ret = Rex::Commands::get( Rex::CMDB::cmdb($cmdb_key) );
+        }
+      }
+
+      if ( !$ret ) {
+
+        # check in global namespace
+        $ret = Rex::Commands::get( Rex::CMDB::cmdb($key) );
+      }
+    }
+  }
+
+  if ( !$ret ) {
+    $ret = $default;
+  }
+
+  # make the variable also known to templates
+  Rex::Commands::set( $key => $ret );
```

Comment:
no problem, keep writing :+1:  it is pretty cool to have people like you.

At this point we don't know which default values we had to set if the parameter is not given via command line or other possibilities.

But the idea of automatically adding known parameters to the template is good. and i'm going to open another task to do this.

I think the problem that the parameters persist in subtask must be solved in another way. Currently we have a task object, and we can add those parameters to this task object. So with every new task we also do have a new object which then is clean.

I'm going to take a look at this.